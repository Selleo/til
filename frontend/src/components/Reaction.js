import React, { useState, useEffect } from 'react'
import { useDispatch } from 'react-redux'
import { handleReaction, checkHasReacted } from '../utils'
import { saveAllPosts, saveAllUsers } from '../store/actions/actions'
import useUser from '../utils/customHooks/useUser'
import classNames from 'classnames'

const Reaction = ({ post, reaction }) => {
  const { id } = post
  const { type, whoReacted, Icon } = reaction
  const [hasReacted, setHasReacted] = useState(false)
  const [reactionNumber, setReactionNumber] = useState(0)
  const user = useUser()
  const dispatch = useDispatch()

  useEffect(() => {
    if (user) {
      const didReact = checkHasReacted(whoReacted, user.uuid)

      setHasReacted(didReact)
      setReactionNumber(whoReacted.length)
    }
  }, [user, whoReacted])

  const toggleReaction = async () => {
    if (hasReacted) {
      await handleReaction(id, 'DELETE', type)
        .then(() => {
          setHasReacted(!hasReacted)
          setReactionNumber(reactionNumber - 1)
          dispatch(saveAllPosts())
          dispatch(saveAllUsers())
        })
        .catch(err => console.error(err))
    } else {
      await handleReaction(id, 'POST', type)
        .then(() => {
          setHasReacted(!hasReacted)
          setReactionNumber(reactionNumber + 1)
          dispatch(saveAllPosts())
          dispatch(saveAllUsers())
        })
        .catch(err => console.error(err))
    }
  }

  const iconActiveClasses = classNames(type, {
    '-fill': hasReacted,
  })

  return (
    <div className="post__single-reaction" onClick={toggleReaction}>
      <Icon width="28px" className={iconActiveClasses} />
      <div>{reactionNumber}</div>
    </div>
  )
}

export default Reaction
