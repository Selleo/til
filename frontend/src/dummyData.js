export const initialState = {
  categories: [],
  currentUser: null,
  users: [],
  posts: []
};

export const user = {
  email: "l.walczak@selleo.com",
  firstName: null,
  image:
    "https://lh5.googleusercontent.com/-4na6W1JJVa0/AAAAAAAAAAI/AAAAAAAAAAA/SDwbLxAgbqc/photo.jpg",
  lastName: null,
  uuid: "36528600-e215-49bf-8d5b-e1096b63c876"
};

export const allUsers = [
  {
    email: "l.walczak@selleo.com",
    firstName: null,
    image:
      "https://lh5.googleusercontent.com/-4na6W1JJVa0/AAAAAAAAAAI/AAAAAAAAAAA/SDwbLxAgbqc/photo.jpg",
    lastName: null,
    uuid: "36528600-e215-49bf-8d5b-e1096b63c876"
  },
  {
    email: "maciekPlacek@selleo.com",
    firstName: null,
    image:
      "https://lh5.googleusercontent.com/-4na6W1JJVa0/AAAAAAAAAAIdsdfjdkjdfi/photo.jpg",
    lastName: null,
    uuid: "36528600-e215-49bf-8d5b-dadlfj33343443"
  }
];

export const categories = [
  { id: 1, name: "java" },
  { id: 2, name: "javascript" }
];

export const posts = [
  {
    author: {
      email: "l.walczak@selleo.com",
      firstName: null,
      image:
        "https://lh5.googleusercontent.com/-4na6W1JJVa0/AAAAAAAAAAI/AAAAAAAAAAA/SDwbLxAgbqc/photo.jpg",
      lastName: null,
      uuid: "36528600-e215-49bf-8d5b-e1096b63c876"
    },
    body: "adsfdf",
    categoriesIds: [2, 1],
    id: 1,
    isPublic: false,
    likes: [
      {
        post_id: 1,
        user_uuid: "36528600-e215-49bf-8d5b-e1096b63c876"
      }
    ],
    likesCount: 1,
    title: "title "
  },
  {
    author: {
      email: "l.walczak@selleo.com",
      firstName: null,
      image:
        "https://lh5.googleusercontent.com/-4na6W1JJVa0/AAAAAAAAAAI/AAAAAAAAAAA/SDwbLxAgbqc/photo.jpg",
      lastName: null,
      uuid: "36528600-e215-49bf-8d5b-e1096b63c876"
    },
    body: "adfefsddf",
    categoriesIds: [3],
    id: 2,
    isPublic: false,
    likes: [],
    likesCount: 0,
    title: "maciekdfd"
  }
];
