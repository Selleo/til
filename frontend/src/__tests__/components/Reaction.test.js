import React from "react";
import { Provider } from "react-redux";
import { render } from "@testing-library/react";
import configureMockStore from "redux-mock-store";
import Reaction from "../../components/Reaction";
import { initialState, posts, user } from "../../dummyData";

describe("<Reaction/>", () => {
  const reaction = posts[0].reactions[0];
  const props = { post: posts[0], reaction };
  const mockStore = configureMockStore([]);
  const store = mockStore({ ...initialState, currentUser: user, posts });

  it("renders properly", () => {
    const { asFragment } = render(
      <Provider store={store}>
        <Reaction {...props} />
      </Provider>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
