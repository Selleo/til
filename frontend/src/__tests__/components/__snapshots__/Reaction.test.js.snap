// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Reaction/> renders properly 1`] = `
<DocumentFragment>
  <p>
    like: 1
  </p>
</DocumentFragment>
`;
